### PR TITLE
cmd: add exit commands

### LINKED ISSUE #2848: Spike on generic exits
# 🎯 Problem to be solved

We need to do a technical spike to figure out possible approaches to support generic exits (see [PRD](https://docs.google.com/document/d/140kknAtq2MWIWPUMIH4odoHtO0PhnHcP3EjMuAVhwU0/edit)).

# 🛠️ Proposed solution

Spike to see if we can craft a generic solution to the problem of exiting distributed validators. The output of this spike should be a design doc containing the research findings.

Note that we also have a [design doc](https://docs.google.com/document/d/1CWnqeuPTgb67igGdEo83Ca7TX8lSCak1Ide_4zI6pio/edit#heading=h.mssb0822zapb) that could be useful for this spike.

# 🧪 Tests

None

### PR DESCRIPTION
**Pull request by gsora**
This PR adds exit-related commands under the `exit` subcommand:

 - `submit-partial-exit`, which signs and submit to an instance of Obol API a partial exit for a given DV in a given cluster lock
 - `broadcast`, which downloads a full exit from an instance of Obol API for a given validator if available, and broadcasts it to the configured beacon node
 - `active-validator-list`, which returns the list of validators which are `ACTIVE_ONGOING` contained in the specified cluster lock (useful for scripting).

Moved `obolapi` mock implementation to `testutil/obolapimock` so other tests can use it.

Added a few utility functions in `eth2util/keystore`, taken from `lido-dv-exit`: since it depends on Charon, we can migrate them.

category: feature
ticket: #2848 

Closes #2848.

### CODE REVIEW

**Review by dB2510 — APPROVED** [review]

**Review by pinebit — APPROVED** [review]

**Review by xenowits — APPROVED** [review]

**Review by pinebit — APPROVED** [review]

File: cmd/exit.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+// sigDataForExit returns the hash tree root for the given exit message, at the given exit epoch.
+func sigDataForExit(ctx context.Context, exit eth2p0.VoluntaryExit, eth2Cl eth2wrap.Client, exitEpoch eth2p0.Epoch) ([32]byte, error) {
+	sigRoot, err := exit.HashTreeRoot()
+	if err != nil {
+		return [32]byte{}, errors.Wrap(err, "exit hash tree root")
+	}
+
+	domain, err := signing.GetDomain(ctx, eth2Cl, signing.DomainExit, exitEpoch)
+	if err != nil {
+		return [32]byte{}, errors.Wrap(err, "get domain")
```
[the remark this answers is not in the document]
Comment (gsora) [reply cmd/exit.go:162]: Considering https://github.com/ObolNetwork/charon/blob/main/docs/goguidelines.md#error-handling, errors here seems okay?

File: cmd/exit_list.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	manifestFilePath := filepath.Join(config.DataDir, "cluster-manifest.pb")
+
+	cl, err := loadClusterManifest(manifestFilePath, lockFilePath)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not load cluster data")
+	}
+
+	eth2Cl, err := eth2Client(ctx, config.BeaconNodeURL)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create eth2 client for specified beacon node")
```
[the remark this answers is not in the document]
Comment (gsora) [reply cmd/exit_list.go:79]: Not necessarily, creating a client for a BN is an opaque procedure, might be more than not being able to connect to it.

File: cmd/exit_fullexit_internal_test.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+				require.NoError(t, err)
+				defer func() {
+					require.NoError(t, beaconMock.Close())
+				}()
+				bnURL = beaconMock.Address()
+			}
+
+			oapiURL := badStr
+			if !test.badOAPIURL {
+				oapiURL = "https://api.obol.tech"
```
[the remark this answers is not in the document]
Comment (gsora) [reply cmd/exit_fullexit_internal_test.go]: That test is just checking for proper CLI arguments parsing, in fact if you look at the flow it requires the test to fail with a specific error.

File: cmd/exit_list.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	return nil
+}
+
+func listActiveVals(ctx context.Context, config exitConfig) ([]string, error) {
+	lockFilePath := filepath.Join(config.DataDir, "cluster-lock.json")
+	manifestFilePath := filepath.Join(config.DataDir, "cluster-manifest.pb")
+
+	cl, err := loadClusterManifest(manifestFilePath, lockFilePath)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not load cluster data")
```
[the remark this answers is not in the document]
Comment (gsora) [reply cmd/exit_list.go:74]: Agreed!

Edited everywhere it applies.

File: cmd/exit_submitpartial.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	valEth2, err := validator.ToETH2()
+	if err != nil {
+		return errors.Wrap(err, "cannot convert validator pubkey to bytes")
+	}
+
+	ctx = log.WithCtx(ctx, z.Str("validator", validator.String()))
+
+	shareIdx, err := keystore.ShareIdxForCluster(cl, *identityKey.PubKey())
+	if err != nil {
+		return errors.Wrap(err, "could not load share index from cluster lock")
```
[the remark this answers is not in the document]
Comment (gsora) [reply cmd/exit_submitpartial.go]: Might be a malformed ENR as well. Will implement this.

File: cmd/exit_submitpartial.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	}
+
+	valKeys, err := rawValKeys.SequencedKeys()
+	if err != nil {
+		return errors.Wrap(err, "could not load keystore")
+	}
+
+	shares, err := keystore.KeysharesToValidatorPubkey(cl, valKeys)
+	if err != nil {
+		return errors.Wrap(err, "could not match keyshares with their counterparty in cluster manifest")
```
[the remark this answers is not in the document]
Comment (gsora) [reply cmd/exit_submitpartial.go]: Ehhh it goes through it but to the user it's still a cluster lock. Implemented.

File: cmd/exit.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (OisinKyne) [reply cmd/exit.go:162]: It is inline with that I suppose. But as I dwell on how we can reduce the customer support load on us, my bias would be towards us trying to make our error messages slightly more verbose and ideally actionable (potentially via a link if we want). Resolving for now but something to consider.

File: eth2util/keystore/keystore.go [hunk 7]
```diff
@@ -225,3 +240,86 @@ func checkDir(dir string) error {
...
+				Share: shares[shareIdx],
+				Index: shareIdx + 1,
+			}
+			found = true
+
+			break
+		}
+
+		if !found {
+			return nil, errors.New("share from provided private key shares slice not found in provided manifest")
```
Comment (OisinKyne) [inline eth2util/keystore/keystore.go:286]:
```suggestion
			return nil, errors.New("public key share from provided private key share not found in provided lock")
```

**Review by OisinKyne — APPROVED** [review]
Not all suggestions are musts. There might be some slight contradictions between them, use best judgement :)

File: eth2util/keystore/keystore.go [hunk 8]
```diff
@@ -225,3 +240,86 @@ func checkDir(dir string) error {
 
 	return nil
 }
+
+// KeysharesToValidatorPubkey maps each share in cl to the associated validator private key.
+// It returns an error if a keyshare does not appear in cl, or if there's a validator public key associated to no
```
[the remark this answers is not in the document]
Comment (gsora) [reply eth2util/keystore/keystore.go:245]: Hmm not really, this function maps a partial key share to its full validator pubkey by looking into the cluster lock, not by doing aggregation or any operation with the private key.

**Comment by gsora** [comment]
Seems to me we fixed all wording issues and the code looks okay.

Merging! 🎉

**Pull request merged**